File: Week 0/OneCounter/Code.jl
function generateBinaryString(n, max) # Strings of binary numbers from 00... to 11... of length n
    return [string(i, base = 2, pad=n) for i in 0:max]
end

function OneCounter(string)        # returns number of ones in decimal
    count = 0
    for i in string
        if (i == '1')
          count += 1  
        end
    end
    return count
end

n = 8
max = 2^n - 1
BinaryStrings = generateBinaryString(n, max)

open("TRACEFILE.txt","w") do file
    for i in BinaryStrings
        count = OneCounter(i)                    # Function returns number of ones in decimal
        c = string(count, base = 2, pad = 4)     # Convert count to binary
        M = "1111"                               # Mask bit
        write(file,"$i $c $M\n")                 # Writing in file
    end
end

File: Week 2/Multiplier/Code.py
import numpy as np
import sympy as sp

n = 4
Max = 2**n

def generateBinaryStrings(n, Max):
    return [bin(i)[2:].zfill(n) for i in range(Max)]

BinaryStrings = generateBinaryStrings(n, Max)

s = "1011"

int(s,2)

def multiplyIntegers(a,b,n, Max):
    return a * b;

numberOfOutputBits = 8

with open("Tracefile.txt", "w") as file: 
    for i in range(Max):
        for j in range(Max):
            I1 = BinaryStrings[i]
            I2 = BinaryStrings[j]
            output = multiplyIntegers(i,j,n, Max)
            mask = '1'*numberOfOutputBits
            o = bin(output)[2:].zfill(numberOfOutputBits)
            str = "{}{} {} {}\n".format(I1, I2, o, mask)
            file.write(str)